src/raftstore/store/snap.rs

```diff
@@ -61,223 +108,519 @@ impl Display for SnapKey {
-    /// the end of file.
-    pub fn save_with_checksum(&mut self) -> io::Result<()> {
-        self.save_impl(true)
+        fn save_impl(&mut self, append_checksum: bool) -> io::Result<()> {
+            debug!("saving to {}", self.file.as_path().display());
+            if let Some((mut f, path)) = self.tmp_file.take() {
+                if append_checksum {
+                    try!(f.write_u32::<BigEndian>(self.digest.as_ref().unwrap().sum32()));
+                }
+                try!(f.flush());
+                let file_len = try!(fs::metadata(&path)).len();
+                let mut size_track = self.size_track.wl();
+                try!(fs::rename(path, self.file.as_path()));
+                *size_track = size_track.saturating_add(file_len);
+            }
+            Ok(())
+        }
     }
 
-    /// Use the content in temporary files replace the target file.
-    ///
-    /// Please note that this method can only be called once.
-    pub fn save(&mut self) -> io::Result<()> {
-        self.save_impl(false)
+    impl SnapshotGenerator for Snap {
+        fn generate(&mut self,
+                    snap: &DbSnapshot,
+                    region: &Region,
+                    snap_data: &mut RaftSnapshotData)
+                    -> raft::Result<()> {
+            try!(self.do_generate(snap, region));
+            let size = try!(self.total_size());
+            snap_data.set_file_size(size);
+            Ok(())
+        }
     }
 
-    fn save_impl(&mut self, append_checksum: bool) -> io::Result<()> {
-        debug!("saving to {}", self.file.as_path().display());
-        if let Some((mut f, path)) = self.tmp_file.take() {
-            if append_checksum {
-                try!(f.write_u32::<BigEndian>(self.digest.sum32()));
+    impl Snap {
+        fn do_generate(&mut self, snap: &DbSnapshot, region: &Region) -> raft::Result<()> {
+            if self.exists() {
+                match self.get_validation_reader().and_then(|r| r.validate()) {
+                    Ok(()) => return Ok(()),
+                    Err(e) => {
+                        error!("[region {}] file {} is invalid, will regenerate: {:?}",
+                               region.get_id(),
+                               self.path(),
+                               e);
+                        try!(self.try_delete());
+                        try!(self.init_to_write());
+                    }
+                }
             }
-            try!(f.flush());
-            let file_len = try!(fs::metadata(&path)).len();
-            let mut size_track = self.size_track.wl();
-            try!(fs::rename(path, self.file.as_path()));
-            *size_track = size_track.saturating_add(file_len);
+
+            let t = Instant::now();
+            let mut snap_size = 0;
+            let mut snap_key_cnt = 0;
+            let (begin_key, end_key) = (enc_start_key(region), enc_end_key(region));
+            for cf in snap.cf_names() {
+                if *cf == *CF_RAFT {
```

Comment:
Please extract to a function, such as need_to_pack, and move blow comments to the function.